crates/sui-indexer-alt-framework/src/ingestion/client.rs

```diff
@@ -214,3 +214,183 @@ impl IngestionClient {
         Ok(Arc::new(data))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use prometheus::Registry;
+    use std::collections::HashMap;
+    use std::sync::Arc;
+    use std::time::Duration;
+    use tokio::sync::Mutex;
+    use tokio_util::bytes::Bytes;
+
+    use crate::ingestion::test_utils::test_checkpoint_data;
+
+    use super::*;
+
+    /// Mock implementation of IngestionClientTrait for testing
+    #[derive(Default)]
+    struct MockIngestionClient {
+        checkpoints: Arc<Mutex<HashMap<u64, FetchData>>>,
+        transient_failures: Arc<Mutex<HashMap<u64, usize>>>,
+        not_found_failures: Arc<Mutex<HashMap<u64, usize>>>,
+    }
+
+    #[async_trait::async_trait]
+    impl IngestionClientTrait for MockIngestionClient {
+        async fn fetch(&self, checkpoint: u64) -> FetchResult {
+            // Check for not found failures
+            if let Some(remaining) = self.not_found_failures.lock().await.get_mut(&checkpoint) {
+                if *remaining > 0 {
+                    *remaining -= 1;
+                    return Err(FetchError::NotFound);
+                }
+            }
+
+            // Check for transient failures
+            if let Some(remaining) = self.transient_failures.lock().await.get_mut(&checkpoint) {
+                if *remaining > 0 {
+                    *remaining -= 1;
+                    return Err(FetchError::Transient {
+                        reason: "mock_transient_error",
+                        error: anyhow::anyhow!("Mock transient error"),
+                    });
+                }
+            }
+
+            // Return the checkpoint data if it exists
+            if let Some(data) = self.checkpoints.lock().await.get(&checkpoint) {
+                Ok(data.clone())
+            } else {
+                Err(FetchError::NotFound)
+            }
```

Comment:
It's fine to write it like this, but in case it's interesting to you, this can also be implemented through method chaining (your preference):

```suggestion
            self
                .checkpoints
                .lock()
                .await
                .get(&checkpoint)
                .cloned()
                .ok_or(FetchError::NotFound)       
```

Using `dashmap`, this would be slightly neater as well:

```rust
            self.checkpoints.get(&checkpoint).cloned().ok_or(FetchError::NotFound)
```